File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
function solution(want, number, discount) {
    let answer = 0;
    const discountMap = new Map();
    const wantMap = new Map();
    want.forEach((elem, idx) => {
        wantMap.set(elem, number[idx]);
    })
    
    // 초기 슬라이딩 윈도우
    for (let i = 0; i < 10; i++) {
        // discountMap.has(discount[i]) ? discountMap.set(discount[i], discountMap.get(discount[i]) + 1) : discountMap.set(discount[i], 1);
        discountMap.set(discount[i], (discountMap.get(discount[i]) || 0) + 1);
    }
    
    // discountMap이 wantMap에 있는 요소를 가지고 있는지 판별
    const isMatching = (map1, map2) => {
        for (let [key, value] of map2) {
            // 원하는 개수보다 할인 중인 개수가 적을 때 (또는 해당 요소가 없을 때)
            if ((map1.get(key) || 0) < value) return false;
        }
        return true;
    }
    
    if (isMatching(discountMap, wantMap)) answer++;
    
    // 슬라이딩 윈도우
    for (let i = 10; i < discount.length; i++) {
        // 맨 앞 요소 빼기
        discountMap.set(discount[i - 10], discountMap.get(discount[i - 10]) - 1);
        // i번째 요소 더하기
        discountMap.set(discount[i], (discountMap.get(discount[i]) || 0) + 1);
        
        // isMatching 함수 돌리기
        if (isMatching(discountMap, wantMap)) answer++;
    }
    
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
// 슬라이딩 윈도우 - 10일에 해당하는 할인 품목 Map 만들기 (discountMap)
// wantMap 만들기
// discountMap에 wantMap의 key와 같은 key가 있는지 확인, 해당 key의 value가 wantMap의 value 이상인지 확인 -> answer++;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        
        Map<String, Integer> wantMap = new HashMap();
        for (int i = 0; i < want.length; i++) {
            wantMap.put(want[i], number[i]);
        }
        
        Map<String, Integer> discountMap = new HashMap();
        for (int i = 0; i < discount.length; i++) {
            // 초기 슬라이딩 윈도우
            discountMap.put(discount[i], discountMap.getOrDefault(discount[i], 0) + 1);
            
            if (i >= 10) {
                discountMap.put(discount[i - 10], discountMap.get(discount[i - 10]) - 1); // 맨 앞 요소 제거
            }
            
            // 슬라이딩 윈도우 크기가 10일 때 discountMap과 wantMap 비교
            if (i >= 9) {
                boolean isMatching = true;  // 회원가입 여부
                for ( String key : wantMap.keySet() ) {
                    // discountMap에서 key 찾기
                    if (discountMap.getOrDefault(key, 0) < wantMap.get(key)) {
                        isMatching = false;
                        break;
                    }
                }
                
                if (isMatching) answer++;
            }
        }
        
        return answer;
    }
}